File: 124-binary-tree-maximum-path-sum/124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        global res
        res = float('-inf')
        def maxGain(node):
            global res 
            if not node:
                return 0
            left = node.left
            right = node.right
            maxleft = max(maxGain(left),0)
            maxright = max(maxGain(right),0)
            res = max(res,max(maxleft,0)+max(maxright,0)+node.val)
            return max(maxleft,maxright)+node.val
        maxGain(root)
        return res
                
        

File: 124-binary-tree-maximum-path-sum/124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxRes = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        processRoot(root);
        return maxRes;
    }
    public int processRoot(TreeNode root){
        if(root.left == null && root.right == null){
            maxRes = Math.max(maxRes, root.val);
            return root.val;
        }
        int left = 0, right = 0;
        if(root.left != null){
            left = Math.max(left,processRoot(root.left));
        }
        if(root.right != null){
            right = Math.max(right,processRoot(root.right));
        }
       // System.out.printf("%d %d %d\n",left, right, root.val);
        maxRes = Math.max(maxRes, left + root.val + right);
        return Math.max(left, right) + root.val;    
    }
}